[ZOrder] Fast approach of interleaving bits

cfmcgrady opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-io/delta/blob/master/CONTRIBUTING.md
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
  6. If applicable, include the corresponding issue number in the PR title and link it in the body.
-->

## Description
to resolve https://github.com/delta-io/delta/pull/1149#discussion_r881782878

Improve interleave bits perf, the basic idea is from http://graphics.stanford.edu/~seander/bithacks.html#InterleaveTableObvious


## How was this patch tested?

<!--
If tests were added, say they were added here. Please make sure to test the changes thoroughly including negative and positive cases if possible.
If the changes were tested in any way other than unit tests, please clarify how you tested step by step (ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future).
If the changes were not tested, please explain why.
-->

Existing UTs.

## Does this PR introduce _any_ user-facing changes?

<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Delta Lake versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.

# Benchmark

Before this PR `spark.databricks.delta.optimize.zorder.fastInterleaveBits.enabled = false`:

```
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-b08 on Mac OS X 10.15.5
[info] Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
[info] 1000000 rows interleave bits benchmark:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] sequence - 1 int columns benchmark                  428            440          16          2.3         427.6       1.0X
[info] sequence - 2 int columns benchmark                  572            605          39          1.7         571.6       0.7X
[info] sequence - 3 int columns benchmark                  749            768          32          1.3         748.6       0.6X
[info] sequence - 4 int columns benchmark                  893            934          64          1.1         892.9       0.5X
[info] random - 1 int columns benchmark                    384            391          11          2.6         383.9       1.1X
[info] random - 2 int columns benchmark                    551            557           7          1.8         550.9       0.8X
[info] random - 3 int columns benchmark                    716            722           6          1.4         716.3       0.6X
[info]  random - 4 int columns benchmark                   922            932          11          1.1         922.4       0.5X
```

After this PR `spark.databricks.delta.optimize.zorder.fastInterleaveBits.enabled  = true`:
```
[info] Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-b08 on Mac OS X 10.15.5
[info] Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
[info] 1000000 rows interleave bits benchmark:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] sequence - 1 int columns benchmark                  299            340          36          3.3         298.5       1.0X
[info] sequence - 2 int columns benchmark                  423            459          50          2.4         423.5       0.7X
[info] sequence - 3 int columns benchmark                  560            579          24          1.8         560.4       0.5X
[info] sequence - 4 int columns benchmark                  694            719          37          1.4         694.0       0.4X
[info] random - 1 int columns benchmark                    270            274           4          3.7         269.8       1.1X
[info] random - 2 int columns benchmark                    418            428          13          2.4         418.4       0.7X
[info] random - 3 int columns benchmark                    548            568          32          1.8         547.6       0.5X
[info]  random - 4 int columns benchmark                   680            686           6          1.5         679.6       0.4X
```

cc and thank the Apache Kyuubi interleave bits author @ulysses-you

cfmcgrady commented:
@vkorukanti @scottsand-db

tdas commented:
@cfmcgrady This is absolutely great addition. I took a very quick cursory look, not in the details of the new logic, but overall PR. My major concern is that this is a lot of new code with very custom logic for handling up to some N columns in a more optimized way, but there isnt a lot of unit tests to really ensure the test coverage of the correctness in these cases. The baseline default case has a simple logic and therefore simple to test. All of these new code paths need sufficient testing. However, it also hard to explicitly conclusively test every combination that have been handled specifically. So here are a bunch suggestions that will improve my confidence in the PR. 

1. Put this new code path behind a flag in DeltaSQLConf so that we can easily turn if off if there are concerns. 

2. For test coverage, lets try to do randomized testing. Generate a sequence of random int, run it through the default and fast path, and make sure that the result is the same. 
  - Make sure that the randomization is based on a randomly picked seed that is printed in the logs, so that if there is an error, we can use the logged seed to reproduce the exact input and output. 
  - Make the number of iterations configurable in the code so that we can run it manually like 1B times (will take a while), but when we check it in, we do it like less iterations that will take say < 10 seconds.
 
what do you think?

cfmcgrady commented:
> Put this new code path behind a flag in DeltaSQLConf so that we can easily turn if off if there are concerns.

Added a new configuration `fast.interleaveBits.enabled`, which is enabled by default.

> For test coverage, lets try to do randomized testing. Generate a sequence of random int, run it through the default and fast path, and make sure that the result is the same.

Added new randomized testing, which runs 100,000,000 random interleaved bit test cases.

```
[info] InterleaveBitsSuite:
[info] - 0 inputs (85 milliseconds)
[info] - 1 input (304 milliseconds)
[info] - 2 inputs (26 milliseconds)
[info] - 3 inputs (30 milliseconds)
[info] - nulls (221 milliseconds)
[info] - consistency (253 milliseconds)
[info] - supported types (11 milliseconds)
[info] - randomization interleave bits (2 minutes, 29 seconds)
```

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/expressions/InterleaveBitsSuite.scala:
```diff
@@ -98,4 +103,26 @@ class InterleaveBitsSuite extends SparkFunSuite with ExpressionEvalHelper {
+
+  test("randomization interleave bits") {
+    val numIters = sys.env
+      .get("NUMBER_OF_ITERATIONS_TO_INTERLEAVE_BITS")
```
where is this env variable set?

cfmcgrady replied:
Currently, we can only set the env variable before we run the unit test.
Something like:
```
NUMBER_OF_ITERATIONS_TO_INTERLEAVE_BITS=20 ./build/sbt "testOnly *InterleaveBitsSuite"
```

Do you have a better idea to make this configurable?

vkorukanti replied:
I think this ok. I wanted to check if there is another change to CI that sets this.

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/expressions/InterleaveBits.scala:
```diff
@@ -67,17 +69,49 @@ case class InterleaveBits(children: Seq[Expression])
+          inputs(0))
+        case 8 => interleave8Ints(inputs(7), inputs(6), inputs(5), inputs(4), inputs(3), inputs(2),
+          inputs(1), inputs(0))
+        case _ => defaultInterleaveBits(inputs, inputs.length)
```
make sure there is a test case for the default case here.

cfmcgrady replied:
added.

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/expressions/InterleaveBitsSuite.scala:
```diff
@@ -98,4 +103,26 @@ class InterleaveBitsSuite extends SparkFunSuite with ExpressionEvalHelper {
     InterleaveBits(Seq(Literal(0.toString))).checkInputDataTypes() != TypeCheckSuccess
   }
+
+  test("randomization interleave bits") {
```
how long does this test take with default 100mil rows?

cfmcgrady replied:
logs from GA
```
[info] InterleaveBitsSuite:
[info] - 0 inputs (85 milliseconds)
[info] - 1 input (304 milliseconds)
[info] - 2 inputs (26 milliseconds)
[info] - 3 inputs (30 milliseconds)
[info] - nulls (221 milliseconds)
[info] - consistency (253 milliseconds)
[info] - supported types (11 milliseconds)
[info] - randomization interleave bits (2 minutes, 29 seconds)
```

Should we set a smaller value?

cfmcgrady commented on core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:
```diff
@@ -690,6 +690,16 @@ trait DeltaSQLConfBase {
+          |When true using the O(4 * 8) algorithm for interleave bit, otherwise the O(32 * n)
+          |algorithm, n is the number of input columns""".stripMargin)
+      .booleanConf
+      .createWithDefault(true)
```
SGTM, set to `false` by default.

cfmcgrady commented on core/src/test/scala/org/apache/spark/sql/delta/expressions/InterleaveBitsBenchmark.scala:
```diff
@@ -0,0 +1,86 @@
+
+  private val numRows = 1 * 1000 * 1000
+
+  private def randomInt(numColumns: Int): Seq[Array[Int]] = {
```
let me add a randomized benchmark later.

vkorukanti reviewed (commented):
LGTM, just one minor comment about reducing the input size to one of the tests.

tdas commented on core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:
```diff
@@ -746,6 +746,16 @@ trait DeltaSQLConfBase {
       .createWithDefault(true)
+
+  val FAST_INTERLEAVE_BITS_ENABLED =
+    buildConf("fast.interleaveBits.enabled")
```
i dont like this name. this interleaving is not really a feature by itself, but part of the larger zorder feature. so i think the name of conf should use the zorder conf naming convention `optimize.zorder.fastInterleaveBits.enabled`

also please colocate this conf with the other zorder confs

cfmcgrady replied:
SGTM, updated.

tdas commented on core/src/test/scala/org/apache/spark/sql/delta/expressions/InterleaveBitsSuite.scala:
```diff
@@ -98,4 +158,26 @@ class InterleaveBitsSuite extends SparkFunSuite with ExpressionEvalHelper {
+      .map(_.toInt)
+      .getOrElse(1000000)
+    var i = 0
+    while (i < numIters) {
```
you are not using a specific random seed. if there is a failure, it will be very hard for any one repro that scenario. So pick a random seed, and print it along with the error message

vkorukanti replied:
The input to the verification failure is already printed with the message.

tdas replied:
okay. lets make sure output with mkString actually prints out the int array correctly.

tdas commented on core/src/test/scala/org/apache/spark/sql/delta/expressions/InterleaveBitsSuite.scala:
```diff
@@ -98,4 +158,26 @@ class InterleaveBitsSuite extends SparkFunSuite with ExpressionEvalHelper {
+      }
+      val r1 = InterleaveBits.interleaveBits(input, true)
+      val r2 = InterleaveBits.interleaveBits(input, false)
+      assert(java.util.Arrays.equals(r1, r2), s"input: ${input.mkString(",")}")
```
is this enough information to debug? how will in the input be printed with mkString? also.. should we print the output arrays as well to compare the bits with and without fast interleaving?

cfmcgrady replied:
> how will in the input be printed with mkString?

It would be something like
```
[info] InterleaveBitsSuite:
[info] - randomization interleave bits *** FAILED *** (32 milliseconds)
[info]   java.util.Arrays.equals(r1, r2) was false input: 1922396969,1778178185,1971203485 (InterleaveBitsSuite.scala:179)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
......
```

cfmcgrady replied:
> should we print the output arrays as well to compare the bits with and without fast interleaving?

Since the results are deterministic, we can easily reproduce them for debugging by running the code `InterleaveBits.interleaveBits(input, false)`

cfmcgrady commented on core/src/main/scala/org/apache/spark/sql/delta/expressions/InterleaveBits.scala:
```diff
@@ -52,9 +53,10 @@ case class InterleaveBits(children: Seq[Expression])
+  private val fastInterleaveBitsEnabled = conf.getConf(DeltaSQLConf.FAST_INTERLEAVE_BITS_ENABLED)
 
   override def eval(input: InternalRow): Any = {
+    val ints = new Array[Int](n)
```
sorry for the late reply.
I'm a little busy with the mid-year wrap-up these days, I will take a look when I'm free, maybe next week.

felipepessoto commented:
> > Put this new code path behind a flag in DeltaSQLConf so that we can easily turn if off if there are concerns.
> 
> Added a new configuration `spark.databricks.delta.fast.interleaveBits.enabled`, which is enabled by default.
> 

@cfmcgrady, The config is actually disabled by default, is that expected?
cc @tdas
